stack-graphs/src/cycles.rs

```diff
@@ -107,34 +111,39 @@ where
         }
     }
 
-    /// Determines whether we should process this path during the path-finding algorithm.  If our
-    /// heuristics decide that this path is a duplicate, or is "non-productive", then we return
-    /// `false`, and the path-finding algorithm will skip this path.
-    pub fn should_process_path<F>(&mut self, path: &P, cmp: F) -> bool
+    /// Determines whether we should process this path during the path-finding algorithm.  If we have seen
+    /// a path with the same start and end node, and the same pre- and postcondition, then we return false.
+    /// Otherwise, we return true.
+    pub fn has_similar_path<Eq>(
+        &mut self,
+        graph: &StackGraph,
+        arena: &mut P::Arena,
+        path: &P,
+        eq: Eq,
+    ) -> bool
     where
-        F: FnMut(&P) -> std::cmp::Ordering,
+        Eq: Fn(&mut P::Arena, &P, &P) -> bool,
     {
         let key = path.key();
-        let paths_with_same_nodes = self.paths.entry(key).or_default();
-        let index = match paths_with_same_nodes.binary_search_by(cmp) {
-            // We've already seen this exact path before; no need to process it again.
-            Ok(_) => return false,
-            // Otherwise add it to the list.
-            Err(index) => index,
-        };
 
-        // Count how many paths we've already processed that have the same endpoints and are
-        // "shorter".
-        let similar_path_count = paths_with_same_nodes
-            .iter()
-            .filter(|similar_path| similar_path.is_shorter_than(path))
-            .count();
-        if similar_path_count > MAX_SIMILAR_PATH_COUNT {
-            return false;
+        if graph.incoming_edge_count(key.end_node) <= 1 {
+            // return false;
+        }
```

Comment:
Is the commented out return intended?  If so, I'd suggest either removing the if block complete or adding a comment explaining how/when we'd want to uncomment the return in the future.

Reply:
I am still not convinced that this optimization is safe, so I'm going to remove it.